source/simulator/solver_schemes.cc

```diff
@@ -563,86 +581,110 @@ namespace aspect
+                                          + test_pressure_residual * test_pressure_residual);
+
+                // Determine if the decrease is sufficient.
+                if (test_residual < (1.0 - alpha * lambda) * residual
+                    ||
+                    line_search_iteration >= parameters.max_newton_line_search_iterations
+                    ||
+                    use_picard)
+                  {
+                    pcout << "      Total relative residual after nonlinear iteration " << nonlinear_iteration+1
+                          << ": " << test_residual/initial_residual << ", norm of the rhs: " << test_residual
+                          << ", newton derivative scaling factor: " << newton_handler->get_newton_derivative_scaling_factor() << std::endl;
+                    residual = test_residual;
+                    break;
+                  }
+                else
+                  {
+
+                    pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
+                          << test_residual << " and going to " << (1.0 - alpha * lambda) * residual
+                          << ", relative residual: " << test_residual/initial_residual << std::endl;
+
+                    /**
+                     * The line search step was not sufficient to decrease the residual
+                     * enough, so we take a smaller step to see if it improves the residual.
+                     */
+                    lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                  }
+
+                line_search_iteration++;
+                Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                       ExcMessage ("This tests the while condition. This condition should "
+                                   "actually never be false, because the break statement "
+                                   "above should have caught it."));
               }
-
-            line_search_iteration++;
-            Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
-                   ExcMessage ("This tests the while condition. This condition should "
-                               "actually never be false, because the break statement "
-                               "above should have caught it."));
+            while (line_search_iteration <= parameters.max_newton_line_search_iterations);
+            // The while condition should actually never be false, because the break statement above should have caught it.
           }
-        while (line_search_iteration <= parameters.max_newton_line_search_iterations);
-        // The while condition should actually never be false, because the break statement above should have caught it.
 
 
         if (use_picard == true)
           {
+            // When we are using (defect corrected) Picard, keep the
+            // newton_derivative_scaling_factor at zero.
```

Comment:
I do not see a zero in the lines below. Am I misunderstanding something? If so, please document what I am missing.

Reply:
Ok, I will add to the documentation that the newton_derivative_scaling_factor is actually set above with the help of these two values. I now see that the naming of the variables below is suboptimal...